src/MSBuild/XMake.cs

```diff
@@ -1005,7 +1007,12 @@ internal static bool BuildProject
             GraphBuildOptions graphBuildOptions,
             bool lowPriority,
             string[] inputResultsCaches,
-            string outputResultsCache
+            string outputResultsCache,
+#if FEATURE_GET_COMMANDLINE
```

Comment:
nit: There are quite a few occurrences of this ifdef already and we're adding more - in this PR and in @Forgind's #7213. Do you think it would be worth introducing an abstraction to cover the two cases (with and without `FEATURE_GET_COMMANDLINE`)? Something with no perf impact like a struct that would wrap the `string` or `string[]`.

Reply:
I though about something like this. But it requires more refactoring, which might be addressed in different PR. 
I'd vote for using one, more complex, data type. In particular I'd use `string[]` and in convert `string commandLine` to it as soon as possible. Not sure if this naive approach is feasible though.

Reply:
That sounds good to me. We do combine it back into a string fairly often, which is a little unfortunate but ok.

As an alternative:
https://github.com/dotnet/msbuild/pull/6789

Reply:
Please note that removing `FEATURE_GET_COMMANDLINE` as proposed in #6789 would break `dotnet build` because the CLI directly calls Main:

https://github.com/dotnet/sdk/blob/14b117b7088653b694e16ac2071fcbf634a2a9ab/src/Cli/Microsoft.DotNet.Cli.Utils/MSBuildForwardingAppWithoutLogging.cs#L139

Reply:
Is there any reason you can't have:
public static int Main(string[] args) {
return Main(string.Join(' ', args);
}

...and remove the directive?

Reply:
`FEATURE_GET_COMMANDLINE` makes us get the command line from `Environment.CommandLine` which offers no way of setting it. We would have to wrap it in our own settable prop to be used by the CLI.

Reply:
I'm not sure what happens if you expose two `Main` overloads. If I understand your proposal we would still have `Main()` calling `Environment.CommandLine` and `Main(string[])` for back compat (and dotnet CLI).